src/hotspot/share/logging/logFileStreamOutput.cpp

```diff
@@ -146,6 +163,36 @@ int LogFileStreamOutput::write_internal(const LogDecorations& decorations, const
   return written;
 }
 
+int LogFileStreamOutput::write_internal(const LogDecorations& decorations, const char* msg) {
+  int msg_len = checked_cast<int>(strlen(msg));
+
+  // Do not handle multiline messages if foldmultilines has been specified
+  if (_fold_multilines) return write_internal_line(decorations, msg, msg_len);
+
+  // Handle multiline strings: split the string replacing newlines with terminators,
+  // and then force write_internal_line to print all of them (i.e. not stopping at the
+  // first null but until msg_len bytes are printed)
+  ALLOW_C_FUNCTION(::malloc, char* dupstr = (char*)::malloc((msg_len + 1) * sizeof(char));)
+  if (dupstr == nullptr) {
+    return 0;
+  }
+  ALLOW_C_FUNCTION(::memcpy, ::memcpy(dupstr, msg, msg_len + 1);)
```

Comment:
Why not use `os::strdup`?

Reply:
We already know the length of the string, so why use something that needs to find the length again?

We shouldn't need the `ALLOW_C_FUNCTION` for memcpy however, the rest of the codebase doesn't use it. Raw malloc here is the good choice, as to avoid circular logging.

Reply:
True ... NMT does use UL so I guess we need a rule that UL can not use NMT.